apollo-router/src/warp_http_server_factory.rs

```diff
@@ -75,9 +84,71 @@ impl HttpServerFactory for WarpHttpServerFactory {
                 .map(|cors_configuration| cors_configuration.into_warp_middleware())
                 .unwrap_or_else(|| Cors::builder().build().into_warp_middleware());
 
-            let routes = get_health_request()
-                .or(get_graphql_request_or_redirect(service.clone()))
+            let plugin_routes = plugin_handlers.into_iter().fold(
+                get_health_request().boxed(),
+                move |acc, (plugin_name, custom_handler)| {
+                    let route = warp::get()
+                        .and(warp::path::full())
+                        .and(warp::any().map(move || custom_handler.clone()))
+                        .and(warp::path("plugins"))
+                        .and(warp::path(plugin_name))
+                        .and(warp::host::optional())
+                        .and(warp::body::bytes())
+                        .and_then(
+                            move |path: FullPath,
+                                  handler: Handler,
+                                  authority: Option<Authority>,
+                                  body: Bytes| async move {
+                                let res = handler.oneshot(http_compat::RequestBuilder::new(
```

Comment:
the headers should probably be passed there too. Example: a plugin that would expose a functionality only if we pass the right token in the `Authorization` header